lessons/module-1/js-pure-functions.md

```diff
@@ -0,0 +1,133 @@
+---
+title: "JS: Pure Functions"
+length: 120
+tags: javascript, functions, functional programming
+---
+
+## Learning Goals
+
+* Speak to what makes a function pure and the benefits it can offer
+* Understand what side effects are and why it's important to avoid them
+* Practice transforming impure functions into more pure and predictable functions
+
+## Vocab
+
+- `Pure Function` A function that has no side effects.
+- `Side Effect` It is a state change that can be seen outside of a function call. (*such as changing a variable outside of a function's scope*)
+
+## What is a pure function?
+
+Over the last few weeks, you've begun writing multiple functions that complete different tasks and utilize parameters & arguments in different ways.  These tasks can include generating a random number, returning a new object, manipulating data, or making an update to the DOM.
+
+A **pure function** is a function that is predictable and always returns the same results or output when given the same input.  Another way of describing this is it's a function that has no side effects, (*it does not make any updates to the outside world*).  It is very likely that you have already written some *pure* functions in past projects.  Let's take a look at an example:
+
```

Comment:
I think the "side effect" concept is important enough that we should probably call it out in another section here, in the same way that you did with 'what is a pure function'

i.e. - 'what is a side effect'

I think that students could fall into a trap of thinking 'as long as I don't modify an array outside of my function, then it's pure' but side effects could be _many_ different things in reality.

Reply:
yeah I would agree! Depending on where else we talk about `state`, calling a side-effect a state change could be confusing, but your example of a variable is a good place to start. Maybe we could give some more examples of pushing into a global array, or something to that effect?

Reply:
Aye, thanks for the feedback on that Will!  I can definitely add a more specific section for "side effects" similar to "pure functions"!  I'll have that update pushed up shortly!  

I agree Scott that talking about "state" changes might be a bit confusing this early on in the program.  But yes, I like the example of pushing into a global array.  I tried to give one example of that in Exercise 2, where they are looking at an example of pushing friends into a global array.  Is that similar to what you're thinking of or would having a more complex example be helpful?